### PR TITLE
Do not suggest `#![feature(...)]` if we are in beta or stable channel.

### LINKED ISSUE #23973: Don’t suggest 'help: add #![feature(…)]' when it’s forbidden
Steps to reproduce:

```sh
mkdir foo
cd foo
multirust override beta  # Or obtain Rust beta in some other way
echo 'extern crate core; fn main(){}' > a.rs
echo '#![feature(core)] extern crate core; fn main(){}' > b.rs
```

The output of `rustc a.rs` is:
```rust
a.rs:1:1: 1:19 error: use of unstable library feature 'core'
a.rs:1 extern crate core; fn main(){}
       ^~~~~~~~~~~~~~~~~~
a.rs:1:19: 1:19 help: add #![feature(core)] to the crate attributes to enable
```

`b.rs` does as suggested by the `help` message, but that fails too (as expected):
```rust
b.rs:1:1: 1:18 error: unstable feature
b.rs:1 #![feature(core)] extern crate core; fn main(){}
       ^~~~~~~~~~~~~~~~~
note: this feature may not be used in the beta release channel
error: aborting due to previous error
```

Proposed fix: in the beta and stable channels (i.e. whenever `#![feature(…)]` is forbidden), the `help` message that suggests adding `#![feature(…)]` should not be shown.

(Clean up the test case with `multirust remove-override; cd ..; rm -r foo`.)

CC @brson

### PR DESCRIPTION
**Pull request by pnkfelix**
Do not suggest `#![feature(...)]` if we are in beta or stable channel.

Fix #23973

### CODE REVIEW

**Comment by pnkfelix** [comment]
(i'm still working on a local build to double-check that this work, and does not break any tests that might be over-zealously checking the help output, but I figured I'd get this up in the meantime.)

**Comment by steveklabnik** [comment]
looks good here. maybe a test?

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -409,6 +409,12 @@ impl<'a> Context<'a> {
 
 pub fn emit_feature_err(diag: &SpanHandler, feature: &str, span: Span, explain: &str) {
     diag.span_err(span, explain);
+
+    // #23973: do not suggest `#![feature(...)]` if we are in beta/stable
```
Comment (steveklabnik) [inline src/libsyntax/feature_gate.rs]: do we leave these kind of comments?

**Comment by pnkfelix** [comment]
I do not yet know how to write tests that are only run on beta/stable or likewise only nightly/development... but I'll look into it.  :)

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (pnkfelix) [inline src/libsyntax/feature_gate.rs]: I do.

**Comment by SimonSapin** [comment]
It doesn’t look like `option_env!("CFG_RELEASE_CHANNEL")` is used to determine whether to allow `#![feature(…])`…

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (pnkfelix) [inline src/libsyntax/feature_gate.rs]: (In particular, the detail of *why* we are early returning requires one read beneath the match statement ... I figure a one-line comment with issue number is worthwhile in that context.  If I had done this by moving the `fileline_help` invocations into the `_ => { ... }` arm instead, then I would not have included the comment, and would have treated the code as self-explanatory in that case.  But as soon as I fiddle with `return` or `break` or other abnormal control-transfers, I try to err on the side of leaving more comments.)

Comment (steveklabnik) [inline src/libsyntax/feature_gate.rs]: interesting, I guess it's a different workflow. I tend to reach for `blame`, which then points to this PR. *shrugs*

**Comment by pnkfelix** [comment]
@SimonSapin hmm let me go look at how that is controlled in that case.

**Comment by SimonSapin** [comment]
I think `rustc::session::config::get_unstable_features_setting` should be used instead.

**Comment by pnkfelix** [comment]
@SimonSapin but I cannot access that from `libsyntax` ... :(

**Comment by SimonSapin** [comment]
CC @brson, winner of the `git blame` lottery :)

**Comment by pnkfelix** [comment]
@SimonSapin However, I will be *happy* to cut and paste the first line from `get_unstable_feature_setting` : 
```rust
    let disable_unstable_features = option_env!("CFG_DISABLE_UNSTABLE_FEATURES").is_some();
```

I.e. this is probably the right environment variable to use, rather than inspecting the value of the `CFG_RELEASE_CHANNEL`

**Comment by SimonSapin** [comment]
Hum, yes, that seems right. The bootstrap key thing is probably not relevant since by this time we’ve already decided to emit an error.

**Comment by pnkfelix** [comment]
and its a net win for the code clarity!  :)

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (pnkfelix) [inline src/libsyntax/feature_gate.rs]: (Oh, don't get me wrong; I'm a heavy user of `blame` too, via `M-x vc-annotate`.  But that can require tracing through a chain of history as code gets moved around or inconsequentially revised by intermediate authors.)

**Comment by pnkfelix** [comment]
@brson I don't see an obvious way for me to write a clear regression test for this.  (I could make a `run-make` test that inspects the output of `rustc --version` to determine if we are running on a release tagged as `-nightly` or `-dev`, but that is ... sketchy ...)

**Comment by alexcrichton** [comment]
Could the `get_unstable_features_setting` be moved to libsyntax? It looks like it doesn't have any dependencies on librustc and I agree with @SimonSapin that it'd be nice to reduce the number of places an environment variable is read.

Other than that though r=me, I agree that adding a test for this is probably too difficult.

**Comment by pnkfelix** [comment]
@alexcrichton its probably higher priority for me to figure out how to fix the tests that this breaks ...

**Comment by pnkfelix** [comment]
@brson @alexcrichton I just ripped out the checks for the help in our test suite ... but I don't know whether this is then a net win overall.

In other words: What is the better of two bad options: failing to check in our test suite that the feature names continue to match expectations (i.e. the current state of this PR), or providing a misleading help message to beta users who attempt to use a gated feature (i.e., the current state of the master branch, without this PR, as is)?

**Comment by nikomatsakis** [comment]
@pnkfelix I don't think it's so very important to test that, personally.

**Comment by brson** [comment]
Agree that losing those tests is minor.

**Comment by pnkfelix** [comment]
I am not going to attempt to move `get_unstable_features_setting` in this PR, though I think that I agree that it should be moved.  (Basically if any further work needs to be done on this PR tonight, then someone else, maybe @alexcrichton, should just close this PR and open up a new one.)

**Pull request merged**